File: src/main/java/org/example/commons/api/Repository.java
package org.example.commons.api;

import javax.persistence.metamodel.SingularAttribute;
import java.util.List;

public interface Repository<T, S> {
    List<T> findAll();
    T findById(S id);
    List<T> findByIds(S... ids);
    List<T> findByIds(List<S> ids);
    <X> List<T> findByColumn(SingularAttribute<T, X> column, X... values);
    <X> List<T> findByColumn(SingularAttribute<T, X> column, List<X> values);
    List<T> filter(Filter<T> filter);
    void create(T... entities);
    void create(List<T> entities);
    void update(T... entities);
    void update(List<T> entities);
    void delete(T... entities);
    void delete(List<T> entities);
}

File: src/main/kotlin/org/example/commons/api/Repository.kt
package org.example.commons.api

import javax.persistence.metamodel.SingularAttribute

interface Repository<T, S> {
    fun findAll(): MutableList<T>
    fun findById(id: S): T
    fun findByIds(vararg ids: S): MutableList<T>
    fun findByIds(ids: Iterable<S>): MutableList<T>
    fun <X> findByColumn(column: SingularAttribute<T, X>, vararg values: X): MutableList<T>
    fun <X> findByColumn(column: SingularAttribute<T, X>, values: Iterable<X>): MutableList<T>
    fun create(vararg entities: T)
    fun create(entities: Iterable<T>)
    fun update(vararg entities: T)
    fun update(entities: Iterable<T>)
    fun delete(vararg entities: T)
    fun delete(entities: Iterable<T>)
}